### PR TITLE
First version

### LINKED ISSUE #1: First version: create underwater object plugin
Create underwater object plugin. Guide: https://github.com/uuvsimulator/uuv_simulator/blob/master/uuv_gazebo_plugins/uuv_gazebo_plugins/src/UnderwaterObjectPlugin.cc
Investigate if already exists some library for that functionality.

### PR DESCRIPTION
**Pull request by mateus-amarante**
The first version containing the UnderwaterObject plugin. I limited to just port it to ros2 and Gazebo 11 plus some adjustments to comply with lint tests.

Notes about migration to ros2:
- Use ament_cmake tools to build and install the library
- Define lint and build tests targeting the Foxy distribution

Notes about migration to gazebo 11
- Replace [ignition::math::Box](https://github.com/uuvsimulator/uuv_simulator/blob/bfb40cb153684a0703173117b6bbf4258e8e71c5/uuv_gazebo_plugins/uuv_gazebo_plugins/include/uuv_gazebo_plugins/BuoyantObject.hh#L77) by [ignition::math::AxisAlignedBox](https://ignitionrobotics.org/api/math/6.2/classignition_1_1math_1_1AxisAlignedBox.html)

Notes about linting:
- I kept the original code (with gazebo-style) and applied `ament_uncrustify`. The result is quite similar.
- Changed the HydrodynamicModel::IDENTIFIER from std::string to const char[] to comply with ament_cpplint
- Rename *.hh *.cc to *.hpp and *.cc to make ament_uncrustify to be compatible with ament_cpplint

Resolve #1

### CODE REVIEW

File: test/underwater_sphere_spawn.test.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    def test_node_names(self, proc_info):
+        result = subprocess.run(['gz', 'topic', '-l'], capture_output=True)
+
+        self.assertEqual(result.returncode, 0)
+
+        topics = result.stdout.decode().split('\n')
+        print(topics)
+
+        self.assertIn('hydrodynamics/current_velocity', topics)
+        self.assertIn('/gazebo/default/sphere/base_link/wrench', topics)
```
Comment (BorgesJVT) [inline test/underwater_sphere_spawn.test.py:69]: Are these topics publishing something? I ran colcon test and also manually, but got no output by the topics.

Comment (BorgesJVT) [reply test/underwater_sphere_spawn.test.py:69]: The `hydrodynamics/current_velocity` topic is published by the _uuv2_gazebo_world_plugins_ package so we can consider a zero input initially. I am just missing an output data.

**Review by BorgesJVT — COMMENTED** [review]
The work is great! 
I think we can quickly finish this first PR.
I think for now would be just to have some output data like _wrench_ or _velocity_.
If I'm not wrong, those are the main parameters in the original package: https://github.com/uuvsimulator/uuv_simulator/blob/master/uuv_gazebo_plugins/uuv_gazebo_plugins/src/UnderwaterObjectPlugin.cc#L212

File: .github/workflows/action-ros-ci.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+on: [pull_request]
+jobs:
+  build:
+    name: build and test
+    runs-on: ubuntu-latest
+    container:
+      image: ubuntu:focal
+    steps:
+      - uses: ros-tooling/setup-ros@0.2.1
+      - uses: ros-tooling/action-ros-ci@0.2.1
```
[the remark this answers is not in the document]
Comment (mateus-amarante) [reply .github/workflows/action-ros-ci.yml:11]: The current setup is working

File: test/underwater_sphere_spawn.test.py (hunk 1 above)
Comment (mateus-amarante) [reply test/underwater_sphere_spawn.test.py:69]: In this test, I'm only checking if the topic exists, which minimally checks that the plugin was loaded correctly. It is expected to have no data being published on this topic. As you said, the world plugin publishes it. This model plugin only subscribes to it.

**Comment by mateus-amarante** [comment]
> We can use our own namespace, but yes, gazebo is necessary: https://github.com/orise-robotics/uuv2_gazebo_world_plugins/blob/port_underwater_current_plugin/include/uuv2_gazebo_world_plugins/UnderwaterCurrentPlugin.hh#L31
>
> Besides this package is planned to work with gazebo, I think it is interesting to have our own namespace since it can be an abstraction for other packages based on other simulators like ignition, for example.

Fixed in 7a64cca

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+find_package(ignition-math6 REQUIRED)
+
+# BUILD
+add_library(underwater_object_plugin SHARED
+  src/BuoyantObject.cpp
+  src/HydrodynamicModel.cpp
+  src/UnderwaterObjectPlugin.cpp
+)
+
+target_include_directories(underwater_object_plugin PUBLIC include)
```
[the remark this answers is not in the document]
Comment (mateus-amarante) [reply CMakeLists.txt:15]: I followed the [ament_cmake user documentation](https://docs.ros.org/en/foxy/Guides/Ament-CMake-Documentation.html)

File: include/uuv2_gazebo_model_plugins/HydrodynamicModel.hpp [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+  virtual void ApplyHydrodynamicForces(
+    double time,
+    const ignition::math::Vector3d & _flowVelWorld);
+
+  /// \brief Computes the added-mass Coriolis matrix Ca.
+
+protected:
+  void ComputeAddedCoriolisMatrix(
+    const Eigen::Vector6d & _vel,
+    const Eigen::Matrix6d & _Ma,
```
[the remark this answers is not in the document]
Comment (mateus-amarante) [reply include/uuv2_gazebo_model_plugins/HydrodynamicModel.hpp:266]: Created #4 to map style improvements

File: include/uuv2_gazebo_model_plugins/HydrodynamicModel.hpp [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+    Eigen::Matrix6d & _Ca) const;
```
[the remark this answers is not in the document]
Comment (mateus-amarante) [reply include/uuv2_gazebo_model_plugins/HydrodynamicModel.hpp:267]: Created #4 to map style improvements

File: include/uuv2_gazebo_model_plugins/HydrodynamicModel.hpp [hunk 6]
```diff
@@ -0,0 +1,589 @@
...
+  /// \brief Linear damping matrix proportional only to the forward speed
+  /// (useful for modeling AUVs). From [1], according to Newman (1977), there
+  /// is a damping force component that linearly increases with the presence
+  /// of forward speed, particularly so for slender bodies.
+  ///
+  /// References:
+  /// [1] Refsnes - 2007 - Nonlinear model-based control of slender body AUVs
+
+protected:
+  Eigen::Matrix6d DLinForwardSpeed;
```
[the remark this answers is not in the document]
Comment (mateus-amarante) [reply include/uuv2_gazebo_model_plugins/HydrodynamicModel.hpp:340]: Created #4 to map style improvements

File: include/uuv2_gazebo_model_plugins/BuoyantObject.hpp [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+protected:
+  physics::LinkPtr link;
+
+  /// \brief If true, the restoring force will be equal to the gravitational
+  // force
+
+protected:
+  bool neutrallyBuoyant;
+
+  // \brief Metacentric width of the robot, used only for surface vessels and
```
[the remark this answers is not in the document]
Comment (mateus-amarante) [reply include/uuv2_gazebo_model_plugins/BuoyantObject.hpp:205]: Created #4 to map style improvements

**Pull request merged**